### PR TITLE
[MBL-720] Configure Remote Config

### PR DESCRIPTION
**Pull request by scottkicks**
# 📲 What

Configures a new `RemoteConfigClient` responsible for loading RC default values, fetching/activating values from the cloud, and providing helpers for determining specific key values.

# 🤔 Why

Remote Config is the feature flagging tool that we are replacing Optimizely with.

# 🛠 How

1. Once FirebaseApp.configure() is called in AppDelegate, we set a `RemoteConfigClient` instance in AppEnvironment to be used throughout the app.
2. On init the new client:
   - loads default values that we may want to change in the future
   - fetches any new values from the cloud. You can actually find these stored in a cached holding pattern on your device
   - "activates" those fetched values. When this happens, it applies those fetched values on top of our existing default values
3. RemoteConfigClient also helps us query our RC values. The way it works is that it gives us either a value from the cloud, if it found one, or a default value based on the provided key.

I've set up 2 keys in our [Kickstarter-Alpha firebase project](https://console.firebase.google.com/u/0/project/kickstarter-ios-alpha-94cf4/config) for testing. Changes to those will reflect in our Feature Flag screen in real time once I get to that bit.

### **Important note regarding the configuration crashes I was seeing during this setup**

In order for RemoteConfig to be initialized the following has to be true:
1. The firebase instance _has to be configured first_
2. Only one copy of firebase can exist in an app. Since we're configuring firebase in AppDelegate we import Firebase into the main target. I tried to keep our new RemoteConfigClient in the Library target, but that requires importing Firebase there as well. However, since the, main target was already importing Library, and thus firebase again, a dependency conflict occurred showing the following error in the console `Class FIRApp is implemented in both...` causing a crash. I needed to keep RemoteConfigClient in the main target to get around this.

    Google intentionally does this to keep multiple copies of Firebase from existing in one app. 
    See [this doc on the Firebase iOS SDK GitHub](https://github.com/firebase/firebase-ios-sdk/blob/master/docs/firebase_in_libraries.md) for more details.

# ✅ Acceptance criteria

- [ ] The should get the correct values from our [firebase alpha project](https://console.firebase.google.com/u/0/project/kickstarter-ios-alpha-94cf4/config) when RemoteConfigClient.fetchCloudValues is called.
- [ ] Consent Management Dialog and Facebook Interstitial behave accordingly based on the RC value

Use the `RemoteConfigFeature+Helpers` to verify

# ⏰ TODO

- Updating the feature flag debug menu and user defaults will be done in the [next ticket](https://kickstarter.atlassian.net/browse/MBL-721)

### CODE REVIEW

File: GoogleService-Info.plist
Comment (scottkicks) [inline GoogleService-Info.plist:1]: I deleted and redownloaded/added this file as I was debugging. It should be the same file and its just for the Alpha build.

File: Library/AppEnvironment.swift
Comment (msadoon) [inline Library/AppEnvironment.swift:250]: remember to update `pushEnvironment` too (only way I knew was looking at the old `optimizelyClient`) in this file on an older branch.

File: Kickstarter.xcodeproj/project.pbxproj
Comment (msadoon) [reply Kickstarter.xcodeproj/project.pbxproj:389]: If you need to start on a new branch, try reverting these changes locally (`git reset --soft commtHash`) and applying the stash to a new branch without the file adds/deletes.

File: Kickstarter-iOS/Library/RemoteConfigClient.swift
Comment (msadoon) [inline Kickstarter-iOS/Library/RemoteConfigClient.swift:1]: So I think we don't need this file.

Here's my thinking: 

The pattern we followed for `OptimizelyClient` is to create it in `AppDelegate` and then conform a protocol for functions we need from it. It was `OptimizelyClientType`.

Then all we do is pass around that `OptimizelyClientType` instance via `AppEnvironment`. 

This is why the real `OptimizelyClient` could easily be replaced with the mock version when we removed the library.

With the new `RemoteConfigClientType` we already conform it to whatever methods we get access to in `RemoteConfig.remoteConfig()`and pass around the `RemoteConfigClientType`, so there is no need for this wrapper class around `RemoteConfig.remoteConfig()` singleton. Just like there was no need for a wrapper around `OptimizelyClient` from `Optimizely` framework.

It's also easier to mock - for that matter add a `MockRemoteConfigClient` in the same way we had `MockOptimizelyClientType`.

File: Kickstarter-iOS/AppDelegate.swift
Comment (msadoon) [inline Kickstarter-iOS/AppDelegate.swift:192]: So my top level pr review comment will address the changes here.

**Review by msadoon — CHANGES_REQUESTED** [review]
Okay lots of writing here, hehe, please bear with:

So generally pretty good start to setting this up. It took me a while to get a grasp of the different ways in which we could setup `RemoteConfig` and I found that there's basically three ways:

1. [Fetch and activate on load](https://firebase.google.com/docs/remote-config/loading#strategy_1_fetch_and_activate_on_load)
2. [Activate behing a loading screen](https://firebase.google.com/docs/remote-config/loading#strategy_2_activate_behind_loading_screen)
3. [Load new values for startup](https://firebase.google.com/docs/remote-config/loading#strategy_3_load_new_values_for_next_startup)

The tutorial was doing no. 2, but because we don't have a loading screen I suggest no. 3.

1. `AppDelegate` `configureFirebase` output should call a new function `configureRemoteConfig` instead of directly calling `AppEnvironment.updateRemoteConfigClient(RemoteConfigClient())`...because there are other steps here for sending configuration signals into the view model.

2. In that function configure `RemoteConfig` in this way:

```
func configureRemoteConfig() {
	let remoteConfig = RemoteConfigClient()

	AppEnvironment.updateRemoteConfigClient(remoteConfig)

    let appDefaults: [String: Any?] = [
      RemoteConfigFeature.consentManagementDialogEnabled.rawValue: false,
      RemoteConfigFeature.facebookLoginInterstitialEnabled.rawValue: false
    ]
    RemoteConfig.remoteConfig().setDefaults(appDefaults as? [String: NSObject])

	RemoteConfig.remoteConfig().activate { _, error in
		guard let remoteConfigActivationError = error else { 
			print("🔮 Remote Config SDK Successfully Activated")

			// TODO: Eventually replace self.viewModel.inputs.didUpdateOptimizelyClient(optimizelyClient), when we remove `MockOptimizelyClient`
			return 
		}

		print("🔴 Remote Config SDK Activation Failed with Error: \(remoteConfigActivationError.localizedDescription)")

		Crashlytics.crashlytics().record(error: remoteConfigActivationError)

		self.viewModel.inputs.optimizelyClientConfigurationFailed()
		// we'll rename this later but keep it this way for now - because notifications are attached to it.
	}

	RemoteConfig.remoteConfig().fetch { }

	self.remoteConfigUpdateListener = RemoteConfig.remoteConfig().addOnConfigUpdateListener { }
	// This will be a global private var in `AppDelegate`, remember to deallocate when `applicationWillTerminate` delegate is called - https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1623111-applicationwillterminate. (New method + signal to and out of view model)
}
```

So how did I know to set it up this way? I looked at the previous Optimizely [setup](https://github.com/kickstarter/ios-oss/pull/1810/files#diff-2b5cc9e2481b28948ced1e33f97217f053db9fd9c2de5070819b45ff03463369L394).

So just doing the reading for no. 3. I understood this setup to be what they meant.
It might not work this way..but I thought I'd suggest because it seemed to be the fastest, least complicated and most similar to the way Optimizely loaded the FFs.

Two cents on testing:
Load the app and see that the features keys on Firebase are accurately reflected. 
(ie. campaign tab is showing, stripe payment sheets are showing) - those are easiest to test because they are large UI things.

Then go ahead and turn those features off on the dashboard. Continue to test the active app and switch between features. Those features should not be hidden at this point. Even backgrounding/foregrounding shouldn't hide them.

Third, kill and relaunch the app. Our dashboard flags should be reflectedand those features should be hidden this gives us confidence our`addOnConfigUpdateListener` cached the changes last time the app was active.

Let me know your thoughts!

File: Library/RemoteConfigClientType.swift
Comment (msadoon) [inline Library/RemoteConfigClientType.swift:3]: Good job using this pattern!

File: Kickstarter-iOS/AppDelegate.swift
Comment (scottkicks) [inline Kickstarter-iOS/AppDelegate.swift:464]: @msadoon this was throwing a compile error saying that `addOnConfigUpdateListener` doesn't exist on the RemoteConfig.remoteConfig() instance. I see it in the docs, but I commented out for now.

**Comment by scottkicks** [comment]
@msadoon this latest commit makes the suggested changes. However, this doesn't compile yet, I just wanted to get it up so you can take a look.

The error I'm getting is that  `RemoteConfig does not conform to expected type RemoteConfigClientType`. I tried to add all of the methods that I think are "required" into this type, but I'm not sure what I'm missing yet.

File: Kickstarter-iOS/AppDelegate.swift
Comment (msadoon) [inline Kickstarter-iOS/AppDelegate.swift:464]: Can remove these lines - because we just need to call `fetch`. no need to error handle or anything..from my reading the firebase instructions it seems like a trigger for the `addOnConfiguUpdateListener`

File: Kickstarter-iOS/AppDelegate.swift
Comment (msadoon) [inline Kickstarter-iOS/AppDelegate.swift:436]: Remove these lines - those Firebase instructions indicate on step 2:

```
 Real-time updates bypass the minimum fetch interval setting.
```

File: Kickstarter-iOS/AppDelegate.swift
Comment (msadoon) [inline Kickstarter-iOS/AppDelegate.swift:459]: Add a `TODO`

File: GoogleService-Info.plist
Comment (msadoon) [reply GoogleService-Info.plist:1]: Ok so let's go ahead and remove this file in this pr.

File: Kickstarter.xcodeproj/project.pbxproj
Comment (msadoon) [inline Kickstarter.xcodeproj/project.pbxproj:10637]: Shouldn't see this...we're using the existing Firebase dependency for now.

File: Library/RemoteConfigClientType.swift
Comment (msadoon) [inline Library/RemoteConfigClientType.swift:16]: Can remove because we don't use/need the error.
We can add this back in later if required.

File: Library/MockRemoteConfigClient.swift
Comment (msadoon) [inline Library/MockRemoteConfigClient.swift:10]: Again we don't need this error - there's no place to test it. So remove it.

File: Library/RemoteConfigClientType.swift
Comment (msadoon) [inline Library/RemoteConfigClientType.swift:32]: Because we aren't tied to the `bool(forKey:` naming structure - let's go ahead and use our own naming structure `isFeatureEnabled` made more sense imo.

File: Kickstarter-iOS/AppDelegate.swift
Comment (msadoon) [inline Kickstarter-iOS/AppDelegate.swift:469]: So two things - 
1. We need to test that this function in-app for those "next launch" steps.
2. We need to find a way to call it through `RemoteConfigClientType`protocol.

File: Library/MockRemoteConfigClient.swift
Comment (msadoon) [inline Library/MockRemoteConfigClient.swift:35]: These are just stubbed functions, but just noting here that the next step is to test them in the view controllers' that use them - probably why tests are failing at the moment.

**Comment by msadoon** [comment]
@scottkicks made some additional comments - were making some progress here.
Let's try to see if this strategy (update on next launch) will work by running build as-is on this branch and testing the scenario described above.

File: Kickstarter.xcodeproj/project.pbxproj
Comment (scottkicks) [reply Kickstarter.xcodeproj/project.pbxproj:10637]: So the firebase package is only added to the main target. In order for these new files that have been added to the Library target to access remote config (` RemoteConfig.remoteConfig()`) we need to add FirebaseRemoteConfig to the same target

File: Kickstarter-iOS/AppDelegate.swift
Comment (msadoon) [reply Kickstarter-iOS/AppDelegate.swift]: Let's discuss in our 1-1

**Comment by scottkicks** [comment]
@msadoon i've made the changes we worked through today. A couple of issues still coming up...

1. RemoteConfig is still not conforming to RemoteConfigClientType even after the package upgrade
2. I started to update `didUpdateOptimizelyClient`, but it started leading me down a path that quickly became complicated. So I've left that commented out for now. It looks for an argument of OptimizelyClientType so it's not as simple as we first thought

File: Kickstarter-iOS/AppDelegate.swift
Comment (msadoon) [inline Kickstarter-iOS/AppDelegate.swift:455]: We'll discuss how to change these in a 1-1 Monday (`line 446` and `line 455`)
Should be pretty easy, we can pair on it.

File: Kickstarter.xcodeproj/project.pbxproj
Comment (msadoon) [inline Kickstarter.xcodeproj/project.pbxproj:1]: Yea so I went ahead and made some of those "Link Binary With Library" changes from the xcode upgrade here, so we can merge this in first and minimize conflicts in that branch.

File: Kickstarter.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved
Comment (msadoon) [inline Kickstarter.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved:1]: Again just changes related to updating ReactiveExtensions and Prelude in SPM

**Comment by nativeksr** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Library/RemoteConfig/RemoteConfigClientType.swift
Comment (msadoon) [inline Library/RemoteConfig/RemoteConfigClientType.swift:4]: This was your protocol - the reason it wasn't mapping to the `RemoteConfig.remoteConfig()` object was because that was ObjC. And - I had to look this up - there's quite a bit of annoying bridging that needs to be done when conforming ObjC to Swift code even if its' a 1-1 protocol. Which leads me to...

File: Library/RemoteConfig/RemoteConfligClient.swift
Comment (msadoon) [inline Library/RemoteConfig/RemoteConfligClient.swift:3]: ...this file! Which is simply a wrapper class around our `RemoteConfig.remoteConfig()` ObjC object.
Notice we hide the actual object as a `private var` and then expose the functions through our protocol.
FYI: this works because its' a _Swift_ class not an _ObjC_ class.

So let's go ahead and test this class with out `MockRemoteConfigClient`.

**Review by msadoon — CHANGES_REQUESTED** [review]
Ok, almost good to go!
So I thought it might faster if I merged changes to update the protocol conformance issue.
It's tested and working as expected - I even got the consent management dialog after setting the dashboard flag to true.
Only 2 things remain before merging:
1. Use the `MockRemoteConfigClient` to write tests for `RemoteConfigClient`
2. Let's handle those input signals for Optimizely setup.

We'll talk monday and wrap this one up eod!

File: Library/RemoteConfig/RemoteConfligClient.swift
Comment (scottkicks) [reply Library/RemoteConfig/RemoteConfligClient.swift:3]: ohh makes total sense!

**Review by msadoon — APPROVED** [review]
Good to go - if tests pass.
When this ends up going to qa party there is another step related to turning on Realtime Config (which I didn't yet go through with you for Production, but is done on Alpha).

**Pull request merged**
